docs/secrets.md

```diff
@@ -0,0 +1,52 @@
+Notification endpoints will require credentials in order to communicate with the various external services.
+
+We do not want users to include those credentials in plain text into Flux scripts. We need to provide a mechanism for users load secrets from a safe storage mechanism and provide those secrets to the endpoints.
+
+**Backing Store:** : There are many backing store possibilities, which need not be discussed here.  Any store that provides secure access for key/value lookup may be used by providing an implementation of the KeyLookup  interface in the flux repository.  
+
+**Fine Grained Authorization**:  keys will generally not be accessible in the flux language.  A secret-enabled function will accept meta-data about the secret (a key) and use it to look up the stored value in memory. The lookup will error if the key is not found or if the user is not authorized to use that key.  
+
+# **Design**
+
+## Internal Functions
+A function that requires secret lookup must: 
+
+- Be built-in to the language
+- Be secrets-aware, meaning that it can acquire the injected secret service, scan its input for secret references, and look up the values for the given secret keys.  
+- Not return secrets under any circumstances (e.g. in error messages) 
+
+The external interface for secret lookup is: 
+```Go
+type SecretService interface {
+// LoadSecret retrieves the secret value v found at key k 
+// ctx is the current query context
+    LoadSecret(ctx context.Context,k string) (string, error)
+}
+```
+
+the internal function must be "secret aware" which means that it knows how to search its own input parameters for references to secrets, so that it can make the proper key-for-value substitution at the appropriate time (as close to when the key is used as possible)
+
+for example, a http.post function may reference a token in the header object by having a sub-object with the form  `{secretKey: <string>}`
+http.post(header: {"Content Type": "application-json", "ApplicationSecret": {secretKey: "FLUX_HTTP_POST_KEY"}}, url: "http://some/url/", data: json_enc)
```

Comment:
Something that occurs to me while I was looking at `http.post` is how would this work when it requires appending or prepending another string to the input?

The following will likely be common.

```
import "influxdata/influxdb/secrets"
token = secrets.get("influxdb_auth_token")
http.post(header: {"Authorization": "Token ${token}"})
```

It makes sense that string interpolation could be used here, but how would this end up working? There has to be some way to evaluate the string expression with secret substitution within the `http.post` function because of the requirements in the specification and I don't see this scenario covered in the document.

My original implementation was going to be just to check if the value was a `values.String` or `values.Object` and perform the secret substitution in that way, but I don't think it will be enough because my method doesn't cover the above.

I think this is possible but it also depends on how string interpolation is implemented as we might need to make string expressions lazily evaluated and the ability to evaluate a string expression with the secret service in certain contexts.

Reply:
We talked on slack about potentially making it so `secrets.get` returned a function and that we could have something that evaluated functions within a secret-aware context. We'll have to flesh it out, but I'm going to approve the current method and we'll experiment with the secret aware stuff since string interpolation will very likely be needed in some capacity.

Reply:
Could this be achieved by creating a `SecretStore` implementation that is namespaced / scoped by some configured secret? This would be before wiring it into flux. So in OSS it could have wide scope. Then in cloud the SecretStore could be user / org scoped?